pkg/controllers/nsxserviceaccount/nsxserviceaccount_controller.go

```diff
@@ -0,0 +1,241 @@
+				return false
+			},
+		}).
+		WithOptions(
+			controller.Options{
+				MaxConcurrentReconciles: runtime.NumCPU(),
+			}).
+		Complete(r)
+}
+
+// Start setup manager and launch GC
+func (r *NSXServiceAccountReconciler) Start(mgr ctrl.Manager) error {
+	err := r.setupWithManager(mgr)
+	if err != nil {
+		return err
+	}
+
+	go r.GarbageCollector(make(chan bool), servicecommon.GCInterval)
+	return nil
+}
+
+// GarbageCollector collect NSXServiceAccount which has been removed from crd.
+// cancel is used to break the loop during UT
+func (r *NSXServiceAccountReconciler) GarbageCollector(cancel chan bool, timeout time.Duration) {
+	ctx := context.Background()
+	log.Info("garbage collector started")
+	for {
+		select {
+		case <-cancel:
+			return
+		case <-time.After(timeout):
+		}
+		nsxServiceAccountUIDSet := r.Service.ListNSXServiceAccountRealization()
+		if len(nsxServiceAccountUIDSet) == 0 {
+			continue
+		}
+		nsxServiceAccountList := &nsxvmwarecomv1alpha1.NSXServiceAccountList{}
+		err := r.Client.List(ctx, nsxServiceAccountList)
+		if err != nil {
+			log.Error(err, "failed to list NSXServiceAccount CR")
+			continue
+		}
+		r.garbageCollector(nsxServiceAccountUIDSet, nsxServiceAccountList)
+	}
+}
+
+func (r *NSXServiceAccountReconciler) garbageCollector(nsxServiceAccountUIDSet sets.String, nsxServiceAccountList *nsxvmwarecomv1alpha1.NSXServiceAccountList) (gcSuccessCount, gcErrorCount uint32) {
+	nsxServiceAccountCRUIDMap := map[string]types.NamespacedName{}
+	for _, nsxServiceAccount := range nsxServiceAccountList.Items {
+		nsxServiceAccountCRUIDMap[string(nsxServiceAccount.UID)] = types.NamespacedName{
+			Namespace: nsxServiceAccount.Namespace,
+			Name:      nsxServiceAccount.Name,
+		}
+	}
+
+	for nsxServiceAccountUID := range nsxServiceAccountUIDSet {
+		if _, ok := nsxServiceAccountCRUIDMap[nsxServiceAccountUID]; ok {
+			continue
+		}
+		log.V(1).Info("GC collected NSXServiceAccount CR", "UID", nsxServiceAccountUID)
+		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, MetricResType)
+		namespacedName := r.Service.GetNSXServiceAccountNameByUID(nsxServiceAccountUID)
+		if namespacedName.Namespace == "" || namespacedName.Name == "" {
+			continue
+		}
+		err := r.Service.DeleteNSXServiceAccount(context.TODO(), nsxServiceAccountCRUIDMap[nsxServiceAccountUID])
+		if err != nil {
+			gcSuccessCount++
+			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteFailTotal, MetricResType)
+		} else {
+			gcErrorCount++
+			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteSuccessTotal, MetricResType)
+		}
+	}
+	return
+}
+
+func (r *NSXServiceAccountReconciler) updateNSXServiceAccountStatus(ctx *context.Context, o *nsxvmwarecomv1alpha1.NSXServiceAccount, e *error) {
+	obj := o
+	if e != nil && *e != nil {
+		obj = o.DeepCopy()
+		obj.Status.Phase = nsxvmwarecomv1alpha1.NSXServiceAccountPhaseFailed
+		obj.Status.Reason = obj.Status.Reason + fmt.Sprintf("Error: %v", *e)
```

Comment:
Will this make Status.Reason longer and longer? Maybe we can keep just the latest reason.

Reply:
Sure. We only set Reason here thus keeping latest reason makes sense.